docs/okeps/okep-5233-pre-assigned-pod-net-id.md

```diff
@@ -0,0 +1,378 @@
+part of the route advertisement feature. The proposal is to have it available all the time. 
+
+
+With `k8s.ovn.org/primary-udn-ipamclaim` being deprecated in favor of the `IPAMClaimReference` field
+in the `NetworkSelectionElement` we have to define the expected behavior. To avoid conflicting 
+settings when `v1.multus-cni.io/default-network` is set the `k8s.ovn.org/primary-udn-ipamclaim` is 
+going to be ignored, it will be reflected in the opposite scenario for backwards compatibility 
+with a plan to remove it in a future release.
+Deprecation plan for the `k8s.ovn.org/primary-udn-ipamclaim` annotation:
+- release-N - emit a warning event stating that the annotation is deprecated and will be removed in a future release.
+- release-N+1 - fail to configure pods with the annotation set.
+- release-N+2 - remove any code handling the annotation, effectively ignoring it.
+
+Note that `GatewayRequest` is not listed, the default gateway is an attribute of the network is not going to be
+configurable per pod.
+
+### Address allocation
+
+OVN-Kubernetes currently [generates](https://github.com/ovn-kubernetes/ovn-kubernetes/blob/3ef29b9a32b04b7917a0afd6b0e9651d17242ed7/go-controller/pkg/util/net.go#L100-L113)
+the overlay MAC addresses from the IPs:
+
+- IPv4: It takes the four octets of the address (e.g `AA.BB.CC.DD`) and uses them to
+create the MAC address with a constant prefix (e.g. `0A:58:AA:BB:CC:DD`).
+- IPv6: Computes a SHA256 checksum from the IPv6 string and uses the first four bytes for the MAC
+address with the `0A:58` constant prefix(e.g. `0A:58:SHA[0]:SHA[1]:SHA[2]:SHA[3]`).
+
+Although unlikely, we need to implement logic that ensures that the MAC address requested through
+the `NetworkSelectionElement` does not conflict with any other configured address on the UDN
+(including addresses consumed by OVN-Kubernetes).
+A similar approach is required for IP address conflict detection.
+When a conflict is detected the pod should not start and an appropriate event should be emited.
+
+When the `NetworkSelectionElement` contains an `IPAMClaimReference` the referenced IPAMClaim should
+reflect the IP allocation status including error reporting through the newly introduced
+`Conditions` status field.
+In the opposite scenario where the `NetworkSelectionElement` does not specify the `IPAMClaimReference`
+the IP allocation is not persisted when the pod is removed.
+
+### Testing Details
+
+
+The following scenarios should be covered in testing:
+
+* VM workloads import into OVN-Kubernetes with no changes to the instances network configuration.
+* Imported VM workloads can live-migrate to another node without any additional traffic disruption.
+* 'v1.multus-cni.io/default-network' cannot be changed after the pod was created.
+* It should be possible to configure the pods MAC or the IP address without configuring the other.
+* Detect MAC and IP address conflicts between the requested addresses for a newly created pods and the addresses that
```

Comment:
to avoid mac address collisions, the only thing I can think of is create a mac allocator in ovnk and track them. Persist them in annotation, and resync the allocator on restart of ovnk cluster manager. Do you have a different idea? If so, how do you propose avoiding mac conflicts? We should probably explicitly state how we plan to do this in the enhancement.

Reply:
The MAC address is already a part of the pod annotation, I have added a plan for the MAC collision detection.